File: heap/minHeap.ts
class MinHeap{
  
  constructor(public data: number[] = [], public length = 0){}

  add(v: number){
    this.data[this.length] = v
    this.heapUp(this.length)
    this.length++
  }

  delete():number | undefined{
    if(this.length === 0) return
    const current = this.data[--this.length]
    const out = this.data.pop() 
    this.data[0] = current
    this.heapDown(0)
    return out
  }

  heapUp(i: number):void{
    if (this.length < 1){
      return
    }
    const parentI = this.getParent(i)
    const partentV = this.data[parentI]
    const currentV = this.data[i]
    if (currentV < partentV){
      this.data[i] = partentV
      this.data[parentI] = currentV
      this.heapUp(parentI)
    }
  }

  heapDown(i: number):void{
    const lChildInd = this.getLChild(i)
    const rChildInd = this.getRChild(i)
    if (lChildInd >= this.length){
      return
    }
    const lChildValue = this.data[lChildInd]
    const rChildValue = this.data[rChildInd]
    const currentValue = this.data[i]
    console.log(lChildValue, rChildValue, currentValue)
    if (lChildValue < rChildValue && lChildValue < currentValue){
      this.data[lChildInd] = currentValue
      this.data[i] = lChildValue
      this.heapDown(lChildInd)
    } else if ( rChildValue < lChildValue && rChildValue < currentValue){
      this.data[rChildInd] = currentValue
      this.data[i] = rChildValue
      this.heapDown(rChildInd)
    }
  }

  getParent(i: number):number {
      return Math.floor((i-1) / 2)
  }

  getRChild(i: number){
    return i * 2 + 2
  }
  
  getLChild(i: number):number{
    return i * 2 + 1
  }
}

// const mh = new MinHeap()

// mh.add(50)
// mh.add(75)
// mh.add(80)
// mh.add(100)
// mh.add(150)
// mh.add(88)
// mh.add(90)
// mh.add(130)
// mh.add(110)
// mh.add(200)
// mh.add(180)
// mh.add(60)
// console.log(mh.data)
// mh.delete()
// console.log(mh.data)


const mh = new MinHeap()

mh.add(100)
mh.add(75)
mh.add(90)
mh.add(60)
mh.add(55)
mh.add(40)
mh.add(78)
mh.add(17)
mh.add(19)

console.log(mh.data)
console.log(mh.data[0])
console.log(mh.delete())

console.log(mh.data)

File: heap/min_heap.py
class MinHeap:

  def __init__(self):
    self.data = {}
    self.length = 0

  def heap_up(self, i):
    if i == 0:
      return
    parent_i = self.get_parent(i)
    parent_v = self.data[parent_i]
    current_v = self.data[i]
    if parent_v > current_v:
      self.data[i], self.data[parent_i] = parent_v, current_v
      self.heap_up(parent_i)

  def heap_down(self, i):
    l_child_i = self.get_l_child(i)
    
    r_child_i = self.get_r_child(i)
    if l_child_i >= self.length:
      return
    if r_child_i == self.length:
      if self.data[l_child_i] < self.data[i]:
        self.data[l_child_i], self.data[i] = self.data[i], self.data[l_child_i]   
      return
    if self.data[l_child_i] < self.data[r_child_i] and self.data[i] > self.data[l_child_i]:
      self.data[l_child_i], self.data[i] = self.data[i], self.data[l_child_i]
      self.heap_down(l_child_i)
    elif self.data[r_child_i] <= self.data[l_child_i] and self.data[i] > self.data[r_child_i]:
      self.data[r_child_i], self.data[i] = self.data[i], self.data[r_child_i]
      self.heap_down(r_child_i) 

  def add(self, v):
    self.data[self.length] = v
    self.heap_up(self.length)
    self.length += 1

  def delete(self):
    if not self.length:
      return
    self.length -= 1
    value = self.data[self.length]
    self.data[self.length] = None
    out = self.data[0]
    self.data[0] = value
    self.heap_down(0)  
    return out

  def get_parent(self, i):
    return (i - 1) // 2

  def get_l_child(self, i):
    return i * 2 + 1

  def get_r_child(self, i):
    return i * 2 + 2

mh = MinHeap()

mh.add(100)
mh.add(75)
mh.add(90)
mh.add(60)
mh.add(55)
mh.add(40)
mh.add(78)
mh.add(17)
mh.add(19)

print(mh.data)
print(mh.data[0])
print(mh.delete())

print(mh.data)
mh.add(58)

print(mh.data)
